sphinx_immaterial/apidoc/cpp/api_parser.py

```diff
@@ -0,0 +1,2217 @@
+    default value.
+    """
+
+    input_content: Optional[bytes] = None
+    """Specifies the content of `.input_path`.
+
+    If unspecified, the content is read from filesystem.
+    """
+
+    compiler_flags: List[str] = dataclasses.field(default_factory=list)
+    """List of compiler flags to pass to Clang."""
+
+    verbose: bool = False
+    """Parse in verbose mode."""
+
+    include_directory_map: Dict[str, str] = dataclasses.field(default_factory=dict)
+    """Maps actual include directories to a displayed directory name.
+
+    The keys should be prefixes of paths specified in error messages/source
+    locations identified by clang.
+
+    The values should be the corresponding prefix to use in the documented
+    :cpp:`#include` paths.
+    """
+
+    allow_paths: List[Pattern] = dataclasses.field(
+        default_factory=lambda: [re.compile("")]
+    )
+    """List of regular expressions matching *allowed* paths.
+
+    Only entities defined in files that match `allowed_paths`, and don't match
+    `disallow_paths`, are documented.  By default all entities are documented,
+    but this default is not normally usable, because it will include entities
+    defined in the standard library and third-party libraries.
+
+    """
+
+    disallow_paths: List[Pattern] = dataclasses.field(default_factory=list)
+    """List of regular expressions matching *disallowed* paths.
+
+    Entities defined in files matching any of these patterns are not documented.
+    """
+
+    disallow_namespaces: List[Pattern] = dataclasses.field(default_factory=list)
+    """List of regular expressions matching *disallowed* namespaces.
+
+    Entities defined in namespaces matching any of the specified patterns are
+    not documented.
+    """
+
+    allow_symbols: List[Pattern] = dataclasses.field(
+        default_factory=lambda: [re.compile("")]
+    )
+    """List of regular expressions matching *allowed* symbols.
+
+    Only symbols matching one of the `.allow_symbols` patterns, and not matching
+    `disallow_symbols`, are documented.  By default, all symbols are allowed.
+    """
+
+    disallow_symbols: List[Pattern] = dataclasses.field(default_factory=list)
+    """List of regular expressions matching *disallowed* symbols.
+
+    Symbols matching any of these patterns are undocumented.
+    """
+
+    allow_macros: List[Pattern] = dataclasses.field(
+        default_factory=lambda: [re.compile("")]
+    )
+    """List of regular expressions matching *allowed* macros.
+
+    Only macros names matching `.allow_macros`, and not matching
+    `.disallow_macros`, are documented.
+    """
+
+    disallow_macros: List[Pattern] = dataclasses.field(default_factory=list)
+    """List of regular expressions matching *disallowed* macro names.
+
+    Macros matching any of these patterns are undocumented.
+    """
+
+    ignore_diagnostics: List[Pattern] = dataclasses.field(default_factory=list)
```

Comment:
I think this could be made an arbitrary bool. In my research, I had a conf.py option to output these diags to the build log (defaulting to false) using the severity of the diag as a logging level. If users really want to use static analysis, then they can use clang-tidy separately.

Reply:
I have found that any compiler errors are usually an indication that the result of parsing will be wrong, sometimes in subtle ways, and should normally not be ignored.  However, for tensorstore I ran into some errors related to an undefined __builtin in a standard library header that I was not able to avoid, and did not seem to cause any problems, so I added this mechanism to suppress them.

Reply:
I know that clang will fail fast if it can't find an included lib, thus no parsing at all. That's why I mentioned the compilation database path (`-p` I think) can be used when third party libs are involved.

I haven't yet looked at how you're normalizing the path relative to the conf.py path, but that might also be a factor when passing certain args to clang.

Reply:
Paths are normalized using `include_directory_map`.  We should probably try to support compilation databases, but one tricky thing about a compilation database is that it doesn't include the compiler built-in include paths.   An alternative is to have the user preprocess the desired headers first using their normal compiler.  The original file/line information is preserved by the added `#line` directives.

Reply:
I think users can specify [`-working-directory`](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-working-directory-arg) to clang. Or, maybe we should provide an option to do it.